hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/MarkerHandler.java

```diff
@@ -0,0 +1,189 @@
+      dispatchingScheduledFuture.cancel(true);
+    }
+    dispatchingExecutorService.shutdown();
+    batchingExecutorService.shutdown();
+  }
+
+  /**
+   * @param markerDir marker directory path
+   * @return all marker paths in the marker directory
+   */
+  public Set<String> getAllMarkers(String markerDir) {
+    MarkerDirState markerDirState = getMarkerDirState(markerDir);
+    return markerDirState.getAllMarkers();
+  }
+
+  /**
+   * @param markerDir marker directory path
+   * @return all marker paths of write IO type "CREATE" and "MERGE"
+   */
+  public Set<String> getCreateAndMergeMarkers(String markerDir) {
+    return getAllMarkers(markerDir).stream()
+        .filter(markerName -> !markerName.endsWith(IOType.APPEND.name()))
+        .collect(Collectors.toSet());
+  }
+
+  /**
+   * @param markerDir  marker directory path
+   * @return {@code true} if the marker directory exists; {@code false} otherwise.
+   */
+  public boolean doesMarkerDirExist(String markerDir) {
+    MarkerDirState markerDirState = getMarkerDirState(markerDir);
+    return markerDirState.exists();
+  }
+
+  /**
+   * Generates a future for an async marker creation request
+   *
+   * The future is added to the marker creation future list and waits for the next batch processing
+   * of marker creation requests.
+   *
+   * @param context Javalin app context
+   * @param markerDir marker directory path
+   * @param markerName marker name
+   * @return the {@code CompletableFuture} instance for the request
+   */
+  public CompletableFuture<String> createMarker(Context context, String markerDir, String markerName) {
+    LOG.info("Request: create marker " + markerDir + " " + markerName);
+    MarkerCreationCompletableFuture future = new MarkerCreationCompletableFuture(context, markerDir, markerName);
+    // Add the future to the list
+    MarkerDirState markerDirState = getMarkerDirState(markerDir);
+    markerDirState.addMarkerCreationFuture(future);
+    if (firstCreationRequestSeen.getAndSet(true)) {
+      dispatchingScheduledFuture = dispatchingExecutorService.scheduleAtFixedRate(markerCreationDispatchingRunnable,
+          timelineServiceConfig.markerBatchIntervalMs, timelineServiceConfig.markerBatchIntervalMs,
+          TimeUnit.MILLISECONDS);
+    }
+    return future;
+  }
+
+  /**
+   * Deletes markers in the directory.
+   *
+   * @param markerDir marker directory path
+   * @return {@code true} if successful; {@code false} otherwise.
+   */
+  public Boolean deleteMarkers(String markerDir) {
+    boolean result = getMarkerDirState(markerDir).deleteAllMarkers();
+    markerDirStateMap.remove(markerDir);
+    return result;
+  }
+
+  private MarkerDirState getMarkerDirState(String markerDir) {
+    MarkerDirState markerDirState = markerDirStateMap.get(markerDir);
+    if (markerDirState == null) {
+      synchronized (markerDirStateMap) {
+        markerDirState = new MarkerDirState(markerDir, timelineServiceConfig.markerBatchNumThreads,
```

Comment:
we should have another if (markerDirState == null) within the sync block.

Reply:
Right, to avoid concurrent instantiation.  Fixed.